File: uis/client/mobile_app/lib/services/trade_finance_service.dart
import 'package:dio/dio.dart';
import 'api_service.dart';

class TradeFinanceService {
  final ApiService _api = ApiService();

  // Letters of Credit  → /trade-finance/*  rewrites to /*  on trade-finance-go
  static const String _lcBase = '/trade-finance/api';
  // Bank Guarantees   → /bank-guarantees/* rewrites to /*  on bank-guarantees-go
  static const String _bgBase = '/bank-guarantees/v1/bank-guarantees';
  // Factoring         → /factoring/*       rewrites to /*  on factoring-go
  static const String _factoringBase = '/factoring/v1/factoring/deals';

  // --- Letters of Credit ---
  Future<Response> getLCs({int page = 1, int limit = 20}) async {
    return await _api.get('$_lcBase/list', queryParameters: {'page': page, 'limit': limit});
  }

  Future<Response> getLCById(String id) async {
    return await _api.get('$_lcBase/get', queryParameters: {'id': id});
  }

  Future<Response> createLC(Map<String, dynamic> data) async {
    return await _api.post('$_lcBase/create', data: data);
  }

  Future<Response> amendLC(String id, Map<String, dynamic> data) async {
    return await _api.post('$_lcBase/update', data: {'id': id, ...data});
  }

  // --- Bank Guarantees ---
  Future<Response> getBankGuarantees({int page = 1, int limit = 20}) async {
    return await _api.get('$_bgBase/list', queryParameters: {'page': page, 'limit': limit});
  }

  Future<Response> getBankGuaranteeById(String id) async {
    return await _api.get('$_bgBase/list', queryParameters: {'id': id});
  }

  Future<Response> createBankGuarantee(Map<String, dynamic> data) async {
    return await _api.post('$_bgBase/create', data: data);
  }

  Future<Response> extendBankGuarantee(String id, Map<String, dynamic> data) async {
    return await _api.post('$_bgBase/update', data: {'id': id, 'action': 'extend', ...data});
  }

  Future<Response> cancelBankGuarantee(String id) async {
    return await _api.post('$_bgBase/update', data: {'id': id, 'action': 'cancel'});
  }

  // --- Factoring ---
  Future<Response> getFactoringApplications({int page = 1, int limit = 20}) async {
    return await _api.get(_factoringBase, queryParameters: {'page': page, 'limit': limit});
  }

  Future<Response> createFactoringApplication(Map<String, dynamic> data) async {
    return await _api.post(_factoringBase, data: data);
  }

  Future<Response> getFactoringInvoices(String id) async {
    return await _api.get('$_factoringBase/$id/invoices');
  }
}

File: uis/client/web2/src/services/trade_finance_service.ts
import { apiService } from './api_service';
import { getErrorMessage } from '../utils/error_utils';
import type { LetterOfCredit, BankGuarantee, FactoringApplication } from '../models/trade_finance';
import { lcFromJson, bgFromJson, factoringFromJson } from '../models/trade_finance';

const LC_BASE = '/trade-finance/api';
const BG_BASE = '/bank-guarantees/v1/bank-guarantees';
const FACTORING_BASE = '/factoring/v1/factoring/deals';

export class TradeFinanceService {
  // ── Letters of Credit ──────────────────────────────────────────────
  async getLCs(page = 1, limit = 20): Promise<LetterOfCredit[]> {
    try {
      const res = await apiService.get(`${LC_BASE}/list`, { params: { page, limit } });
      if (res.status === 200) {
        const raw: any[] = Array.isArray(res.data) ? res.data : (res.data?.data ?? []);
        return raw.map(lcFromJson);
      }
      return [];
    } catch (err) {
      console.error('getLCs failed:', err);
      return [];
    }
  }

  async getLCById(id: string): Promise<LetterOfCredit | null> {
    try {
      const res = await apiService.get(`${LC_BASE}/get`, { params: { id } });
      if (res.status === 200) {
        const raw = res.data?.data ?? res.data;
        return lcFromJson(raw);
      }
      return null;
    } catch (err) {
      console.error('getLCById failed:', err);
      return null;
    }
  }

  async createLC(data: Record<string, any>): Promise<{ success: boolean; message: string }> {
    try {
      const res = await apiService.post(`${LC_BASE}/create`, data);
      if (res.status === 200 || res.status === 201) {
        return { success: true, message: 'Letter of Credit application submitted!' };
      }
      return { success: false, message: getErrorMessage(res.data, 'Failed to submit LC application') };
    } catch (err) {
      return { success: false, message: getErrorMessage(err, 'Failed to submit LC application') };
    }
  }

  async amendLC(id: string, data: Record<string, any>): Promise<{ success: boolean; message: string }> {
    try {
      const res = await apiService.post(`${LC_BASE}/update`, { id, ...data });
      if (res.status === 200 || res.status === 201) {
        return { success: true, message: 'LC amended successfully!' };
      }
      return { success: false, message: getErrorMessage(res.data, 'Failed to amend LC') };
    } catch (err) {
      return { success: false, message: getErrorMessage(err, 'Failed to amend LC') };
    }
  }

  // ── Bank Guarantees ─────────────────────────────────────────────────
  async getBankGuarantees(page = 1, limit = 20): Promise<BankGuarantee[]> {
    try {
      const res = await apiService.get(`${BG_BASE}/list`, { params: { page, limit } });
      if (res.status === 200) {
        const raw: any[] = Array.isArray(res.data) ? res.data : (res.data?.data ?? []);
        return raw.map(bgFromJson);
      }
      return [];
    } catch (err) {
      console.error('getBankGuarantees failed:', err);
      return [];
    }
  }

  async createBankGuarantee(data: Record<string, any>): Promise<{ success: boolean; message: string }> {
    try {
      const res = await apiService.post(`${BG_BASE}/create`, data);
      if (res.status === 200 || res.status === 201) {
        return { success: true, message: 'Bank guarantee request submitted!' };
      }
      return { success: false, message: getErrorMessage(res.data, 'Failed to submit guarantee request') };
    } catch (err) {
      return { success: false, message: getErrorMessage(err, 'Failed to submit guarantee request') };
    }
  }

  async extendBankGuarantee(id: string, data: Record<string, any>): Promise<{ success: boolean; message: string }> {
    try {
      const res = await apiService.post(`${BG_BASE}/update`, { id, action: 'extend', ...data });
      if (res.status === 200 || res.status === 201) {
        return { success: true, message: 'Bank guarantee extended!' };
      }
      return { success: false, message: getErrorMessage(res.data, 'Failed to extend guarantee') };
    } catch (err) {
      return { success: false, message: getErrorMessage(err, 'Failed to extend guarantee') };
    }
  }

  async cancelBankGuarantee(id: string): Promise<{ success: boolean; message: string }> {
    try {
      const res = await apiService.post(`${BG_BASE}/update`, { id, action: 'cancel' });
      if (res.status === 200 || res.status === 201) {
        return { success: true, message: 'Bank guarantee cancelled!' };
      }
      return { success: false, message: getErrorMessage(res.data, 'Failed to cancel guarantee') };
    } catch (err) {
      return { success: false, message: getErrorMessage(err, 'Failed to cancel guarantee') };
    }
  }

  // ── Factoring ────────────────────────────────────────────────────────
  async getFactoringApplications(page = 1, limit = 20): Promise<FactoringApplication[]> {
    try {
      const res = await apiService.get(FACTORING_BASE, { params: { page, limit } });
      if (res.status === 200) {
        const raw: any[] = Array.isArray(res.data) ? res.data : (res.data?.data ?? []);
        return raw.map(factoringFromJson);
      }
      return [];
    } catch (err) {
      console.error('getFactoringApplications failed:', err);
      return [];
    }
  }

  async createFactoringApplication(data: Record<string, any>): Promise<{ success: boolean; message: string }> {
    try {
      const res = await apiService.post(FACTORING_BASE, data);
      if (res.status === 200 || res.status === 201) {
        return { success: true, message: 'Factoring application submitted!' };
      }
      return { success: false, message: getErrorMessage(res.data, 'Failed to submit factoring application') };
    } catch (err) {
      return { success: false, message: getErrorMessage(err, 'Failed to submit factoring application') };
    }
  }

  async getFactoringInvoices(id: string): Promise<any[]> {
    try {
      const res = await apiService.get(`${FACTORING_BASE}/${id}/invoices`);
      if (res.status === 200) {
        return Array.isArray(res.data) ? res.data : (res.data?.data ?? []);
      }
      return [];
    } catch (err) {
      console.error('getFactoringInvoices failed:', err);
      return [];
    }
  }
}

export const tradeFinanceService = new TradeFinanceService();
